### ISSUE TITLE
flask socketio parallelize using redis except nginx(sticky session)

### ISSUE THREAD

**Issue by bridge-ai** [miguelgrinberg/Flask-SocketIO#694]
Hello @miguelgrinberg 

I have only one server and I want to parallelize flask socketio application as much as the CPU I have using gunicorn.
Should I use nginx to turn on the 'sticky session' flag even if I use a message queue server like redis?
I don't want to use nginx for simplicity and I have to use redis anyway.

ps. I prefer gunicorn instead nginx to parallelize my application because I can share quite large data block in memory at once.
My several gunicorn workers share the memory in read only.
I have several rest apis serving same contents with socketio. (These apis are already using redis.)
In other words, I'm serving a contents with rest api and socketio simultaneously.

**refs.**
https://github.com/miguelgrinberg/Flask-SocketIO/issues/643
https://github.com/miguelgrinberg/Flask-SocketIO/issues/402#issuecomment-278683493

**Comment by miguelgrinberg** [comment]
Gunicorn does not support sticky sessions, so it cannot be used to support multiple SocketIO servers at this time. The way I do it is to start several single-worker Gunicorns, and then load balance with nginx configured for sticky sessions. Unfortunately I think your idea of sharing memory across workers is not going to work because of the lack of sticky session support.

**Comment by bridge-ai** [comment]
@miguelgrinberg 

By your answer, is there any reasons what I should do use Gunicorn?
I think I don't use Gunicorn and set up my socketio application based on flask only and serving through nginx.

**Comment by miguelgrinberg** [comment]
Both approaches are equivalent. Single-worker Gunicorn instances are functionally the same as using the web server that comes with Eventlet or Gevent. In both cases nginx will be in front, so you can test both options and use the one that works best for you. I tend to prefer Gunicorn instances, but it's just a preference.

**Comment by bridge-ai** [comment]
May I make flask-socketio works sessionless using redis?
I think flask-socketio can work sessionless and perfectly parallelized if I save session data into redis cache instead each local memory.
How do you think about this idea?
Is this reasonable?

**Comment by miguelgrinberg** [comment]
What do you mean by sessionless? Is this a reference to sticky sessions? If that's what you are asking, no, sticky sessions are required in the current implementation.

You could replace memory storage with Redis and then that would allow you to work without sticky sessions, but performance is going to suffer because of having to constantly access Redis for the state data associated with each customer.

**Comment by bridge-ai** [comment]
Sorry, I missed.
I misunderstood that the flask is already associated with redis after I've seen your video that you refer redis integrated with celery.
Thank you!

**Comment by miguelgrinberg** [comment]
Yes, Redis is already integrated, but it is used for a different purpose, not to store client state.

**Comment by bridge-ai** [comment]
I hope flask-socketio be improved to store even client state to be parallelized easily and independently.
Thank you, I think flask and flask-socketio are awesome!
